### PR TITLE
[vlm, llm] feat: support Kimi K3 LLM and VLM

### PR DESCRIPTION
**Pull request by dyedd**
## Change Summary

- What changed?

  - Added Kimi K3 language-backbone support, including KDA/MLA attention, latent MoE, AttnRes residual handling, and pipeline-boundary support.
  - Added Kimi K3 LLM and VLM model configurations.
  - Added Hugging Face <-> MCore checkpoint conversion mappings for the K3 language model, vision tower, and image projector.
  - Added K3 chat-template support, shared Kimi multimodal plugin naming, and K3 task-encoder integration.
  - Added K3 pretraining, SFT, checkpoint-conversion, and language-only training examples.
  - Added and updated tests for K3 operators, checkpoint normalization, chat templates, examples, dataset validation, and existing Kimi plugin behavior.
  - Updated model-support documentation and GPU dependencies.

- Which issue or requirement does this address?

  - Adds the LoongForge implementation required to train, fine-tune, and convert Kimi K3 LLM and VLM checkpoints.
  - Keeps the existing K2.x multimodal templates and data paths compatible while sharing the common Kimi media-processing implementation.

- Which modules are affected?
  - Kimi K3 foundation model and attention/MoE layers
  - Vision encoder and image projector configuration
  - Multimodal data loading, task encoding, and chat templates
  - Hugging Face/MCore checkpoint conversion
  - Training and checkpoint-conversion examples
  - Model registry, configuration mapping, documentation, and tests
  - GPU dependencies: `transformers==5.12.1`, `fla-core==0.5.2`

## Validation

- [x] I ran the relevant local checks.
- [x] I added or updated tests where appropriate.
- [x] I checked compatibility with existing APIs and configurations.

Validation performed:

- `git diff --check` passed before and after commit.
- Targeted Python 3.11 compilation checks passed for the updated Kimi plugin, task encoder, chat-template, and related test module.
- Existing K2.x Kimi templates remain registered and are covered by the updated plugin tests.
- Full pytest and GPU training validation were not run in the current environment because the required training dependencies were unavailable.

## CI, GPU, and Image Impact

- [x] This change affects dependencies, Dockerfiles, patches, or runtime environments.
- [x] GPU validation is required. Suggested suite: `llm_vlm` / `embodied`.
- [x] A local candidate image should be built during GPU validation.
- [ ] No GPU or image validation is required.

The dependency update should be included in the candidate image before GPU validation.

## Reviewer Notes

- Kimi K3 KDA currently requires `CP=1`; virtual pipeline parallelism is not supported.
- The K3 VLM vision execution path reuses the existing `MoonVisionModel` implementation. Structural checkpoint loading is wired, but independent official-K3 ViT numerical parity still requires validation.
- End-to-end GPU training, checkpoint round-trip, and logits parity should be completed in CI or on a K3-capable GPU environment.

### CODE REVIEW

File: examples/kimi_k3/finetuning/sft_kimi_k3.sh [hunk 1]
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env bash
```
Comment (gsl322) [inline examples/kimi_k3/finetuning/sft_kimi_k3.sh]: The script format shall be consistent with that of other models, and the same applies to other scripts.

File: examples/kimi_k3/checkpoint_convert/convert_kimi_k3_mcore_to_hf_bf16.sh [hunk 2]
```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/env bash
```
Comment (gsl322) [inline examples/kimi_k3/checkpoint_convert/convert_kimi_k3_mcore_to_hf_bf16.sh:1]: There is no FP8 conversion script available.

File: examples/kimi_k3/checkpoint_convert/convert_kimi_k3_hf_to_mcore.sh [hunk 3]
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env bash
+# Copyright 2026 The LoongForge Authors.
```
Comment (gsl322) [inline examples/kimi_k3/checkpoint_convert/convert_kimi_k3_hf_to_mcore.sh:2]: There is no FP8 conversion script available.

File: configs/models/image_projector/kimi_k3_patch_merger.yaml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+_target_: loongforge.models.encoder.KimiK3PatchMergerConfig
+
```
Comment (gsl322) [inline configs/models/image_projector/kimi_k3_patch_merger.yaml:2]: configs/models/image_projector/ckpt_convert/kimi_k3_patch_merger_convert.yaml  and  configs/models/image_projector/kimi_k3_patch_merger.yaml， can the two files  be merged into other yaml files?

File: loongforge/models/foundation/kimi_k3/kimi_k3_ops.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    score_weight = score_norm.weight.float() * score_proj.weight.squeeze(0).float()
+    probabilities = torch.softmax((normalized * score_weight).sum(dim=-1), dim=-1)
+    mixed = (probabilities.unsqueeze(-1) * rows_float).sum(dim=-2).to(rows.dtype)
+    return mixed if output_norm is None else output_norm(mixed)
+
+
+def sum_grads_across_tp(module: nn.Module) -> None:
+    """Mark a module's replicated parameters so TP ranks sum their gradients."""
+    for parameter in module.parameters():
+        parameter.sum_gradients_across_tp_domain = True
```
Comment (gsl322) [inline loongforge/models/foundation/kimi_k3/kimi_k3_ops.py]: parameter.sum_gradients_across_tp_domain = True，  Loong-Megatron finalize_model_grads.py 只识别：average_gradients_across_tp_domain和allreduce_gradients_across_tp_domain，因此 TP > 1 训练时，这些复制参数不会进入 TP gradient all-reduce。

File: examples/kimi_k3/pretrain/pretrain_kimi_k3.sh [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  --bf16 \
+  --load "$CHECKPOINT_PATH" \
+  --save "$CHECKPOINT_SAVE_PATH" \
+  --save-interval "${SAVE_INTERVAL:-100}" \
+  --eval-interval "${EVAL_INTERVAL:-30}" \
+  --eval-iters "${EVAL_ITERS:-10}" \
+  --no-load-optim \
+  --no-load-rng \
+  --moe-grouped-gemm \
+  --tensor-model-parallel-size "${TP:-1}" \
```
Comment (gsl322) [inline examples/kimi_k3/pretrain/pretrain_kimi_k3.sh]: 应该加上--sequence-parallel

File: loongforge/models/foundation/kimi_k3/kimi_k3_ops.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    def __init__(self, config) -> None:
+        super().__init__()
+
+    def forward(self, inputs: torch.Tensor) -> torch.Tensor:
+        """Apply SiTU to a fused gate/up projection."""
+        return situ_and_mul(inputs)
+
+
+class KimiRMSNorm(nn.Module):
+    """Kimi's FP32-accumulating RMS normalization."""
```
Comment (gsl322) [inline loongforge/models/foundation/kimi_k3/kimi_k3_ops.py]: Could we reuse the generic RMSNorm?

File: loongforge/models/foundation/kimi_k3/kimi_k3_ops.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+"""Kimi K3 numerical operators."""
+
+from functools import cache
+
+import torch
+from torch import nn
+
+# K3's SiTU shape parameters (HF: activation_situ_beta, activation_situ_linear_beta).
+SITU_BETA = 4.0
+SITU_LINEAR_BETA = 25.0
```
Comment (gsl322) [inline loongforge/models/foundation/kimi_k3/kimi_k3_ops.py]: Should be specified in config, not hard‑coded.

File: configs/models/image_projector/kimi_k3_patch_merger.yaml (hunk 4 above)
Comment (dyedd) [reply configs/models/image_projector/kimi_k3_patch_merger.yaml:2]: These two K3 YAML files serve different purposes: the `projector `YAML defines the runtime module, while the `*_convert.yaml` file defines the HF-to-MCore checkpoint key mapping. K2.6 follows the same separation through the shared patch_merger_adapter.yaml and `ckpt_convert/patch_merger_adapter_convert.yaml`, which are referenced by kimi_k2_6.yaml rather than named after K2.6. K3 cannot reuse that pair because its projector is structurally different: **it has bias-free linear layers, no pre-norm LayerNorm, and an output RMSNorm. Therefore, K3 needs a distinct runtime preset and checkpoint mapping.**

File: examples/kimi_k3/pretrain/pretrain_kimi_k3.sh [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+  --distributed-timeout-minutes 60
+  --enable-experimental
+)
+
+MOE_ARGS=(
+  --moe-router-topk 16
+  --moe-grouped-gemm
+  --moe-router-enable-expert-bias
+  --moe-router-pre-softmax
+  --moe-router-bias-update-rate 0.0
```
Comment (gsl322) [inline examples/kimi_k3/pretrain/pretrain_kimi_k3.sh:102]: --moe-router-bias-update-rate 0.0?

File: examples/kimi_k3/pretrain/pretrain_kimi_k3.sh [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+  --dataloader-type external
+  --split 99990,8,2
+  --add-question-in-pretrain
+  --enable-discard-sample
+  --num-workers 16
+)
+
+TRAINING_ARGS=(
+  --training-phase pretrain
+  --seq-length 32768
```
Comment (gsl322) [inline examples/kimi_k3/pretrain/pretrain_kimi_k3.sh:69]: pretrain should add --no-create-attention-mask-in-dataloader.

**Review by gsl322 — APPROVED** [review]
